### ISSUE TITLE
Table2 `columnWidths` prop is not working as expected

### ISSUE THREAD

**Issue by N1kto** [palantir/blueprint#5778]
#### Environment
"@blueprintjs/table": "4.7.9"

#### Code Sandbox

Link to a minimal repro: [demo](https://codesandbox.io/s/zealous-johnson-7y0jc5?file=/src/App.js:0-23)

#### Steps to reproduce
Checkout the demo code:
- initially `colWidths` contains `null` value for the 1st column width and 230 for the 2nd column;
- further we have an effect which updates `colWidths`: we set 400 width for the 1st column, 2nd column remains the same;
- finally we have a toggle which allows to set both columns' widths to 300;

Now follow these steps:
1. Load the page
2. Click the "Set all cols 300 width" button 
3. Click "Set all cols 300 width" again

#### Actual behavior

1. Once loaded column widths are: some "default" value for 1st col and 230 for the 2nd one
2. Upon 1st click column widths are: 400 and 230
3. Upon 2nd click column widths are: 300 and 300

#### Expected behavior
1. Once loaded, column width should be 400 and 230 due to effect we ran on mount
2. Upon 1st click on "Set all cols 300 width" rendered columns should have 300 and 300 widths

#### Possible solution
Adding `colWidths` (which is passed via `columnWidths` prop) to `cellRendererDependencies` "fixes" the issue;

#### Additional observations
1. It seems like `Table2` uses _previous_ value of  `columnWidths`, hence lagging the most recent value;
2. If you try scrolling the table, it "forces" table to re-render with proper column widths;

P.s. this separate issue creation was [requested](https://github.com/palantir/blueprint/issues/5264#issuecomment-1332237109) in #5264.

**Comment by N1kto** [comment]
@adidahiya  you can comment out the effect you mentioned. It's not the effect that is called too early (btw is it documented that certain props' updates are not allowed before `onCompleteRender` is fired?), but rather smth else, because when you click "set all cols 300" button the column widths are still not updated... Until you scroll or click the button the 2nd time. In fact, I no longer think that `null` values in initial column widths array are an issue, because even if you set all column widths (as per demo code):
```
const [colWidths, setColWidths] = useState([400, 230]);
```
It's still not working as I'd expect (when changing the `colWidths` via "Set all cols..." button). Or am I misusing the `columndWitdhs` prop? Please advise.

**Comment by rafaellichen** [comment]
@N1kto you can give a key to the table in order to force a re-render when the `colWidths` changes.
```js
const [tableKey, setTableKey] = useState(new Date().getTime())
useEffect(() => {
    setTableKey(new Date().getTime())
}, [colWidths])
...
return <Table2 key={tableKey} />
```

**Comment by N1kto** [comment]
@rafaellichen , yeah I can. But that's not a point of this issue. The point of the issue is unexpected `colWidths` prop behavior. 
And as I mentioned initially, to have the table render with proper widths it's good enough to pass `colWidths` values to `cellRendererDependencies` prop, which, imo, is better than abusing the `key` prop like you suggest.
